### PR TITLE
Handle when same name and same version both project and package are referenced in solution

### LINKED ISSUE #6795: "An item with the same key has already been added" - when project graph has projectName == packageRef name (of same version)
The problem occurs when a dll is referenced as a nuget package and as a project reference.

If there are 3 projects, `Source`, `Intermediate` and `DoublyReferencedTarget`

Then the problem will occur when `Source` has:
- ProjectReference to project `Intermediate`
- ProjectReference to project `DoublyReferencedTarget`

And 'Intermediate' has 
- PackageReference to project `DoublyReferencedTarget`. 

Thus project DoublyReferencedTarget is referenced twice in two different ways, which creates this error.

I realise that this is a genuine error, but a better error message would be very useful. Some of our solutions contains 100's of projects (a bad idea I know), and it's just taken me a week to track down this problem.

I have created a repo to demonstrate the problem at https://github.com/ceddlyburge/nuget-error.

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: [Issue#6795](https://github.com/NuGet/Home/issues/6795)
Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: The problem occurs project and package with same name and same version referenced directly and transitively to project. Let's call them "DoublyReferenced" for both project and package and with version is "1.0.0"

```
   Source  - - - - - ProjectInterMed -> Package(DoublyReferenced) V1.00 (transitive ref)
             \
               \    
                 \ - Project(DoublyReferenced) V1.00 (direct ref)
```
                           

Then it fails with vague error "" in VS:

image

Even on cli with dotnet.exe it fails with another vague error, actually it's telling duplicate same name keys are created and we have conflict.
image

Currently below scenario works, after examination it actual doesn't create key duplicate key as our case, only one key for project created [due to project over package preference](https://github.com/NuGet/Home/issues/6795#issuecomment-740221814). It looks very similar to our case in question but actually it's not. 

```
   Source  - - - - - -> Package(DoublyReferenced) V1.00 (direct ref)
             \
               \    
                 \ - Project(DoublyReferenced) V1.00 (direct ref, winner)
```

Even it works it cause squiggly yellow warning with Roslyn.
image

There're only 2 solutions to this problem. 
1. Try to solve issue by choosing one over other (package vs project). Since [we already have project over package preference ](https://github.com/NuGet/Home/issues/6795#issuecomment-740221814) I currently implemented this approach in my code. 
Still this one have issues: 
  a.  Above Roslyn warning keep continue to show up with or without this fix until we remove one of them.

image

  b. Project reference always winner, it's implicit to user and maybe it's not user actually wanted, maybe yes. All different combinations of direct and transient references works except below 1 case.
  c. If package direct reference and project is transitive reference then VS doesn't even compile
image

image

2. Show explicit warning/error and let user know there are both project and package with same name/version referenced so user can take action. Since main complain of customers was message was too implicit/vague this will definitely solve the pain point.
Maybe this one is most clean and confusion free solution. I can change to implementation, please just give me your feedback.
Also in this case do we need to introduce new warning NUxxxx code?

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:

### CODE REVIEW

**Comment by ericsampson** [comment]
This would be great to have a better error message/handling of this, I just hit it and it took me a while to figure out what was going on. Cheers!

**Comment by erdembayar** [comment]
@NuGet/nuget-client
Do we need to show warning message to user that we picked up project over package in this case instead of doing it silently?

Also I investigated [issue#10368](https://github.com/NuGet/Home/issues/10368) related this but it requires design change, and ROI is very small since it's just uncommon edge case. So want to keep fix as it's. I'm working on adding unit test.

**Comment by nkolev92** [comment]
> Do we need to show warning message to user that we picked up project over package in this case instead of doing it silently?

Imo, not as part of this change. That's not something we want to introduce at this point as no one is really asking for it.

**Comment by erdembayar** [comment]
@NuGet/nuget-client 
Ready for review, please review.

**Comment by nkolev92** [comment]
@erdembayar 

Can you please:

* Fix the title to indicate what the change is, not `Fix {issue title}`. 
* Add a more detailed description about your changes. What will the experience be like now and why you made that decision.

**Comment by erdembayar** [comment]
> @erdembayar
> 
> Can you please:
> 
> * Fix the title to indicate what the change is, not `Fix {issue title}`.
> * Add a more detailed description about your changes. What will the experience be like now and why you made that decision.

It looks I missed one point. How to check which one is more close dependency over other one? Preference project over package is valid if both of them is same level of distant direct/transitive dependencies. But one is closer than other then that one should win over. I think we already have such logic somewhere for packages.

**Comment by erdembayar** [comment]
> @erdembayar
> 
> Can you please:
> 
> * Fix the title to indicate what the change is, not `Fix {issue title}`.
> * Add a more detailed description about your changes. What will the experience be like now and why you made that decision.

Can you check now? Now I more toward to show warning/error direction instead of solve problem by choose one over other (project vs package).  By showing clear warning/error we make user take a action to solve problem for good.
Initial assumption of this is just bug was below direct referencing scenario works. But it's not really working (Roslyn warning) and bit different problem:

```
   Source  - - - - - -> Package(DoublyReferenced) V1.00 (direct ref)
             \
               \    
                 \ - Project(DoublyReferenced) V1.00 (direct ref, winner)
```

**Comment by nkolev92** [comment]
> By showing clear warning/error we make user take a action to solve problem for good.

Project was preferred over Package, but not in all scenarios. 

Fixing this bug just establishes the already known expectation. 
Adding more visibility to projects being preferred over packages is something that's lateral to the current problem.

**Comment by erdembayar** [comment]
> > By showing clear warning/error we make user take a action to solve problem for good.
> 
> Project was preferred over Package, but not in all scenarios.
> 
> Fixing this bug just establishes the already known expectation.
> Adding more visibility to projects being preferred over packages is something that's lateral to the current problem.

Can you recommend next course of action? Is my current fix I submitted ok for bug fix? Do I need to change anything?

**Comment by nkolev92** [comment]
I think the first idea of continuing to choosing project over package is more appropriate. Especially because that matches current behavior.

I'll review the code later.

**Comment by erdembayar** [comment]
> I think the first idea of continuing to choosing project over package is more appropriate. Especially because that matches current behavior.
> 
> I'll review the code later.

@nkolev92 
Please review when you have time.

**Comment by nkolev92** [comment]
@erdembayar Sorry, this slipped through my fingers last week. I've added it to my queue.

**Comment by nkolev92** [comment]
@erdembayar 

Reviewing your description, I think proposal 1 makes sense. 

a. We should file an issue for that. Start it in our repo. The fix might actually end up being on project-system side, but it's NUGet scenario. 
b. I think this is similar to `a`. Personally given that projects have always been preferred, I'd keep it as is for now, just focus on fixing the bug. Adding a warning, if any, can be a different task.
c. That's a very nice catch. However, if you add `PackageVersion` in the DoubleReferencedTarget project, the bug you are fixing doesn't exist and the behavior is the same. We should file an issue on the dotnet/sdk side for this scenario.

It sounds to me like this is what you implemented, so I'll review the code.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 1]
```diff
@@ -283,6 +285,85 @@ public class LockFileBuilder
...
+            foreach (var item in librariesByNameAndVersion)
+            {
+                int count = item.Count();
+
+                if (count == 2)
+                {
+                    var first = item.First();
+                    var second = item.Last();
+
+                    if (first.Type == second.Type)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Can this happen?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 2]
```diff
@@ -148,6 +148,7 @@ public class LockFileBuilder
                 }
             }
 
+            ValidateLockFileLibrary(lockFile.Libraries);
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Given that we're repairing the list after the fact, do we know why we got both here in the first place?

What's the list here when the project has a version that's different from the package.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs (hunk 1 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Not really, I tried to be more cautious, now removing.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs (hunk 2 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: We're coming here because we care about reference `name` and `version`, but not type. 
So having same name references one from project and another from library (with same version) causes dictionary key collision when `name + version` used as key . 

Here I created situation you're interested, project version is `1.0.0` and package version if `2.0.0`:

```
source    ---- > ` DoublyReferencedTarget.csproj Version 1.0.0`
            \
              \ 
                \--> `Intermediate.csproj` -- > `DoublyReferencedTarget" Version="2.0.0` 
```
  
As you can see there is only one `DoublyReferencedTarget` reference here with version `2.0.0` occur:

image

But in case they have same version `1.0.0` then you see 2 `DoublyReferencedTarget` object same version `1.0.0` exist:

```
source    ---- > ` DoublyReferencedTarget.csproj Version 1.0.0`
            \
              \ 
                \--> `Intermediate.csproj` -- > `DoublyReferencedTarget" Version="1.0.0` 
```

image

**Review by erdembayar — COMMENTED** [review]
@nkolev92 
Thank you for your review.
I addressed your PR comments, please check.

**Comment by erdembayar** [comment]
@nkolev92 
Can you review again please?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Thanks for the explanation.  

What do you think about removing this exception check and simply ranking the references?
Consider writing a generic method to avoid typos causing problems. 

We can chat offline if that's easier.

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: @nkolev92 
Thank you for your suggestion. I revised the code, please check. I removed exception check since it's not possible scenario.
But writing generic method is little bit problematic. I can have something like below but problem is `second.Type` and `first.Type` part. Here `LockFileLibrary` and `LockFileTargetLibrary` both of them are already public api, so better not to make any changes to make them generic. Maybe I can use casting or reflection for that line but maybe it's overkill.

```
        private void ValidateLibrary<T, V>(IList<T> libraries, IGrouping<Tuple<string, Versioning.NuGetVersion>, V> item)
        {
            if (item.Count() == 2)
            {
                var first = item.First();
                var second = item.Last();

                // Prefer project reference over package reference, so remove the the package reference.
                libraries.Remove(RankReferences(second.Type) > RankReferences(first.Type) ? second : first);
            }
        }
```

**Comment by erdembayar** [comment]
@nkolev92 
Please review again when you have time.

**Comment by nkolev92** [comment]
@erdembayar 

I'll take a look. 

Consider using the `re-request` review button. 
stuff

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 3]
```diff
@@ -283,6 +284,63 @@ public class LockFileBuilder
             return lockFile;
         }
 
+        private void ValidateLockFileLibrary(IList<LockFileLibrary> libraries)
+        {
+            ILookup<Tuple<string, Versioning.NuGetVersion>, LockFileLibrary> librariesByNameAndVersion = libraries.ToLookup(lib => Tuple.Create(lib.Name, lib.Version));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Cleanup the types. 

Refer to rule #26 - https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Given that this is a hot path and an unlikely occurrence at that, any way we can make it cheaper allocation wise?

**Review by nkolev92 — COMMENTED** [review]
Once you clarified the implementation decisions, I think we're going in the right direction. 

Given that we're in a relatively hot path, I'm wondering if there are ways we can make this cheaper.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs (hunk 3 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: @nkolev92 
Please check current implementation. If you think we need some kind of actual `perf` data then please let me know. I was working on it but something else need attention.

**Comment by ericsampson** [comment]
uhh this probably shouldn't be auto closed, lol

**Comment by erdembayar** [comment]
@nkolev92 
Here is dotnetbenchmark result. I don't see much difference in both run time duration and memory allocation for before and after my change. Actually it performs bit better on edge case we're interested in.

Code used for benchmarking is [here](https://gist.github.com/erdembayar/00f3826804a8da7be54362cb8993eec0).

1. Without any change.

  [Host]     : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT
  DefaultJob : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT


|       Method |     Mean |   Error |  StdDev | Rank |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|------------- |---------:|--------:|--------:|-----:|--------:|--------:|------:|----------:|
| Test0:Normal | 488.6 us | 3.69 us | 3.45 us |    1 | 41.9922 | 10.2539 |     - |    259 KB |

2. With my PR changes. Below `DuplicateId` is edge case which we're interested in.

.NET SDK=6.0.100-preview.5.21302.13
  [Host]     : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT
  DefaultJob : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT


|            Method |     Mean |   Error |  StdDev | Ratio | RatioSD | Rank |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|------------------ |---------:|--------:|--------:|------:|--------:|-----:|--------:|--------:|------:|----------:|
| Test0:DuplicateId | 336.6 us | 2.46 us | 2.05 us |  1.00 |    0.00 |    1 | 28.8086 |  8.3008 |     - |    179 KB |
|      Test0:Normal | 487.1 us | 4.45 us | 4.16 us |  1.45 |    0.02 |    2 | 41.9922 | 10.2539 |     - |    259 KB |

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 4]
```diff
@@ -283,6 +285,79 @@ public class LockFileBuilder
             return lockFile;
         }
 
+        private void ValidateLockFileLibrary(IList<LockFileLibrary> libraries)
+        {
+            var libNameVersion = libraries.Select(l => l.Name + " " + l.Version);
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Would it be at all simpler to just create a dictionary, and iterate over the libs one by one and updating the selected ref as we go?

Currently

* Distinct uses a set internally - 1 pass: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Linq/src/System/Linq/Distinct.cs. 

* The ToLookup statement creates a dict and then it iterates over that, allocating a list there. 
* Finally, iterate over the main list to remove the dups. 

All of that can be replaced with 1 dictionary and 1 loop. 
The moment you find a conflict, rank references and dedup. 

Same applies to both methods.

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: @nkolev92 
Could you be able to check now? I pushed new changes.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 5]
```diff
@@ -283,6 +285,89 @@ public class LockFileBuilder
...
+                        libraryReferences[libraryKey] = library;
+                    }
+                }
+                else
+                {
+                    libraryReferences[libraryKey] = library;
+                }
+            }
+
+            lockFile.Libraries = new List<LockFileLibrary>(libraryReferences.Count);
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Would it make sense to *only* do this when the libraryReferences count is *not* equivalent to the libraries count? 

That means we did something. No need to allocate beyond that.

**Review by nkolev92 — COMMENTED** [review]
One small question.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs (hunk 5 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Sure, thank you for your review. 
I was doing it in initial implementation differently, but missed in new implementation. 
Please check now.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 6]
```diff
@@ -283,6 +285,99 @@ public class LockFileBuilder
             return lockFile;
         }
 
+        private void ValidateLockFileLibrary(LockFile lockFile)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Given that this method's name does not clearly communicate  what it does, consider it to 
`EnsureUniqueLockFileLibraries`

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 7]
```diff
@@ -148,6 +148,7 @@ public class LockFileBuilder
                 }
             }
 
+            ValidateLockFileLibrary(lockFile);
             var libraries = lockFile.Libraries.ToDictionary(lib => Tuple.Create(lib.Name, lib.Version));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: I'm sorry I missed this earlier, do you think it'd be possible to just have the method above return the dictionary instead? That way we have *no* wasted allocations whatsoever, we just create an extra dictionary.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: I renamed to `GetUniqueLockFileLibraries`

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs (hunk 7 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: Great point, I refactored the code. I  should have thought about it before. Please check now.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 8]
```diff
@@ -283,6 +284,99 @@ public class LockFileBuilder
             return lockFile;
         }
 
+        private Dictionary<Tuple<string, NuGetVersion>, LockFileLibrary> GetUniqueLockFileLibraries(LockFile lockFile)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs]: This method is called `GetUniqueLockFileLibraries`, while the other one is `EnsureUniqueLockFileTargetLibraries`, but they really are doing the same thing. 

I'd recommend making the naming pattern match.

**Review by nkolev92 — APPROVED** [review]
Great job!
👏
